file/file.go

```diff
@@ -0,0 +1,166 @@
+package file
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	basicnode "github.com/ipld/go-ipld-prime/node/basic"
+)
+
+// NewUnixFSFile attempts to construct an ipld node from the base protobuf node representing the
+// root of a unixfs File.
+// It provides a `bytes` view over the file, along with access to io.Reader streaming access
+// to file data.
+func NewUnixFSFile(ctx context.Context, substrate ipld.Node, lsys *ipld.LinkSystem) (StreamableByteNode, error) {
+	if substrate.Kind() == ipld.Kind_Bytes {
+		// A raw / single-node file.
+		return &singleNodeFile{substrate, 0}, nil
+	}
+	return &shardNodeFile{ctx, lsys, substrate, nil}, nil
+}
+
+// A StreamableByteNode is an ipld.Node that can be streamed over. It is guaranteed to have a Bytes type.
+type StreamableByteNode interface {
+	ipld.Node
+	io.Reader
+}
+
+type singleNodeFile struct {
+	ipld.Node
+	ptr int
+}
+
+func (f *singleNodeFile) Read(p []byte) (int, error) {
+	buf, err := f.Node.AsBytes()
+	if err != nil {
+		return 0, err
+	}
+	if f.ptr >= len(buf) {
+		return 0, io.EOF
+	}
+	n := copy(p, buf[f.ptr:])
+	f.ptr += n
+	return n, nil
+}
+
+type shardNodeFile struct {
+	ctx       context.Context
+	lsys      *ipld.LinkSystem
+	substrate ipld.Node
+	rdr       io.Reader
+}
+
+var _ ipld.Node = (*shardNodeFile)(nil)
+
+func (s *shardNodeFile) Read(p []byte) (int, error) {
+	// collect the sub-nodes on first use
+	if s.rdr == nil {
+		links, err := s.substrate.LookupByString("Links")
+		if err != nil {
+			return 0, err
+		}
+		readers := make([]io.Reader, 0)
+		lnki := links.ListIterator()
+		for !lnki.Done() {
+			_, lnk, err := lnki.Next()
+			if err != nil {
+				return 0, err
+			}
+			if pbl, ok := lnk.(dagpb.PBLink); ok {
+				target, err := s.lsys.Load(ipld.LinkContext{Ctx: s.ctx}, pbl.Hash.Link(), basicnode.Prototype.Any)
+				if err != nil {
+					return 0, err
+				}
+
+				asFSNode, err := NewUnixFSFile(s.ctx, target, s.lsys)
```

Comment:
An interesting way of recursing.  This will work correctly, but may cause long chains of `Read` calls delegating to `Read` calls delegating to `Read` calls if it's a deeply recursive structure for a large file, right?

(I'd have no objection to merging that as a first pass anyway.  It's probably dominated by IO costs, etc, and it would be something easy to revisit later without external API change if necessary.)

Reply:
it only does it as the read reaches subsequent nodes that need to be accessed, so the tree is loaded as it's accessed.
I can probably do a bit better at deleting references to the multi-reader upon hitting EOF to help the garbage collector clean up better for big files.

Reply:
I'm more worried about keeping the context associated to the New event with the node for it's whole lifetime. I wonder what the right way to manage that context lifetime is.